Source/SwiftyDropbox/Shared/Handwritten/OAuth/OAuth.swift

```diff
@@ -196,39 +200,55 @@ open class DropboxOAuthManager {
         return false
     }
 
-    func extractFromRedirectURL(_ url: URL) -> DropboxOAuthResult {
-        var results = [String: String]()
-        let pairs  = url.fragment?.components(separatedBy: "&") ?? []
-
-        for pair in pairs {
-            let kv = pair.components(separatedBy: "=")
-            results.updateValue(kv[1], forKey: kv[0])
-        }
-
+    func extractFromRedirectURL(_ url: URL, completion: @escaping DropboxOAuthCompletion) {
+        let results = OAuthUtils.extractParamsFromUrl(url)
         if let error = results["error"] {
-            let desc = results["error_description"]?.replacingOccurrences(of: "+", with: " ").removingPercentEncoding
-            if results["error"] != "access_denied" {
-                return .cancel
+            let result: DropboxOAuthResult
+            let desc = results[OAuthConstants.errorDescription]?
+                .replacingOccurrences(of: "+", with: " ")
+                .removingPercentEncoding
+            if results[OAuthConstants.errorKey] != "access_denied" {
```

Comment:
I know this was already here before you came and refactored, but i really dont like how we do all these string comparisons. It'd be nice to have an actual object you can work with. Please see comment in `OauthUtils` regarding changing the return type for `extractParamsFromUrl`.

Reply:
will make a struct that wraps the results. thanks!

Reply:
talked in person, making a new struct maybe a bit too much overhead. I will make sure the code that handles auth results are easier to follow/read, by adding doc strings and better variable naming.